### PR TITLE
Internal `Context` object cleanup and safe "Proxy" access to `ContextBatch`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Clean up the `Context` representation and use the safer "Proxy".

## 🔍 What does this change?

1. `context::Inner` has been merged into `Context` so, `ContextMut` is now wrapping `Context` instead of `Inner`
1. `Context` and `ContextMut` could be converted into each other without any compile time restriction -> so they've now been merged into one
1. It has been made impossible to get direct access to `ContextBatch` from `Context` without using `BatchReadProxy`.
1. Altering the `ContextBatch` is only possible when converting `PreContext` to `Context`

### Does this require a change to the docs?

No, only internal changes

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201707629991362/1201770326087536/f) (_internal_)

### CODE REVIEW

File: packages/engine/src/datastore/table/proxy.rs [hunk 1]
```diff
@@ -62,7 +63,7 @@ impl<K: Batch> Clone for BatchReadProxy<K> {
         // SAFETY: Since a BatchReadProxy already exists, the existing BatchReadProxy
         //         must already have the shared lock, so no writer can currently have
         //         the lock, so it must be possible to take the shared lock again.
-        let locked = unsafe { self.arc.raw() }.try_lock_shared();
+        let locked = unsafe { RwLock::raw(&self.arc).try_lock_shared() };
         assert!(locked, "Clone BatchReadProxy");
```
Comment (Alfred-Mountfield) [inline packages/engine/src/datastore/table/proxy.rs:67]: I know you didn't write it but this seems like a strange message for an assert

File: packages/engine/src/datastore/table/proxy.rs [hunk 2]
```diff
@@ -83,34 +85,36 @@ pub struct BatchWriteProxy<K: Batch> {
...
             Ok(BatchWriteProxy { arc: arc.clone() })
         } else {
             Err(Error::from("Couldn't acquire exclusive lock on object"))
         }
     }
 
-    pub fn inner_mut(&mut self) -> &mut K {
+    pub fn batch(&self) -> &K {
         let ptr = self.arc.data_ptr();
         // Safety: This method uses unsafe code to get a mutable reference to the underlying data.
```
Comment (Alfred-Mountfield) [inline packages/engine/src/datastore/table/proxy.rs:98]: This isn't a mutable reference anymore

File: packages/engine/src/datastore/table/proxy.rs [hunk 3]
```diff
@@ -83,34 +85,36 @@ pub struct BatchWriteProxy<K: Batch> {
...
-        unsafe { &*ptr }
+        // However since this object acts as a guarantee of no other read/write locks existing,
+        // then data races stemming from the dereferencing happening within this method cannot
+        // happen
+        unsafe { &mut *ptr }
     }
 }
 
 impl<K: Batch> Drop for BatchWriteProxy<K> {
+    // SAFETY: We are holding a shared lock, so we can unlock it
```
Comment (Alfred-Mountfield) [inline packages/engine/src/datastore/table/proxy.rs:115]: Is this not holding an exclusive lock at this point?

File: packages/engine/src/datastore/table/proxy.rs [hunk 4]
```diff
@@ -72,7 +73,8 @@ impl<K: Batch> Clone for BatchReadProxy<K> {
 
 impl<K: Batch> Drop for BatchReadProxy<K> {
     fn drop(&mut self) {
-        unsafe { self.arc.raw().unlock_shared() }
+        // SAFETY: We are holding a shared lock, so we can unlock it
```
Comment (Alfred-Mountfield) [inline packages/engine/src/datastore/table/proxy.rs:76]: I don't quite understand this. Is it releasing this specific shared lock? 
Or is it releasing all shared locks?

File: packages/engine/src/datastore/table/proxy.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/engine/src/datastore/table/proxy.rs:98]: Git messed this up. I changed the name from `inner` to `batch` and `inner_mut` to `batch_mut`, but I also changed the order of `batch` and `batch_mut`, so git thinks, `inner_mut` got `batch` and `inner` got `batch_mut`.

File: packages/engine/src/datastore/table/proxy.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/engine/src/datastore/table/proxy.rs:67]: I adjusted the assertion message, good catch!

File: packages/engine/src/datastore/table/proxy.rs (hunk 3 above)
Comment (TimDiekmann) [reply packages/engine/src/datastore/table/proxy.rs:115]: Yes, Copy-Paste error, thanks. I also changed the description to be the same as in `BatchReadProxy::drop`.

File: packages/engine/src/datastore/table/proxy.rs (hunk 4 above)
Comment (TimDiekmann) [reply packages/engine/src/datastore/table/proxy.rs:76]: It's releasing `self`'s lock only. I clarified the comment.

**Review by litvand — CHANGES_REQUESTED** [review]
Nice work overall

File: packages/engine/src/datastore/table/proxy.rs [hunk 5]
```diff
@@ -112,7 +115,7 @@ impl<K: Batch> BatchWriteProxy<K> {
 }
 
 impl<K: Batch> Drop for BatchWriteProxy<K> {
-    // SAFETY: We are holding a shared lock, so we can unlock it
+    // SAFETY: `BatchReadProxy` is guaranteed to have a unique lock, so unlocking is safe.
```
Comment (litvand) [inline packages/engine/src/datastore/table/proxy.rs:118]:
```suggestion
    // SAFETY: `BatchWriteProxy` is guaranteed to have a unique lock, so unlocking is safe.
```

**Review by litvand — CHANGES_REQUESTED** [review]

**Review by litvand — APPROVED** [review]

**Review by litvand — APPROVED** [review]

File: packages/engine/src/datastore/table/proxy.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/engine/src/datastore/table/proxy.rs]: I have rewritten the `SAFETY`-sections in that file.

**Review by litvand — APPROVED** [review]

**Review by Alfred-Mountfield — APPROVED** [review]

**Pull request merged**